internal/query/projection/user_grant.go

```diff
@@ -0,0 +1,187 @@
+					Reduce: p.reduceRemoved,
+				},
+				{
+					Event:  usergrant.UserGrantCascadeRemovedType,
+					Reduce: p.reduceRemoved,
+				},
+				{
+					Event:  usergrant.UserGrantDeactivatedType,
+					Reduce: p.reduceDeactivated,
+				},
+				{
+					Event:  usergrant.UserGrantReactivatedType,
+					Reduce: p.reduceReactivated,
+				},
+			},
+		},
+	}
+}
+
+type UserGrantColumn string
+
+const (
+	UserGrantID            = "id"
+	UserGrantResourceOwner = "resource_owner"
+	UserGrantCreationDate  = "creation_date"
+	UserGrantChangeDate    = "change_date"
+	UserGrantSequence      = "sequence"
+	UserGrantUserID        = "user_id"
+	UserGrantProjectID     = "project_id"
+	UserGrantGrantID       = "grant_id"
+	UserGrantRoles         = "roles"
+	UserGrantState         = "state"
+)
+
+func (p *UserGrantProjection) reduceAdded(event eventstore.Event) (*handler.Statement, error) {
+	e, ok := event.(*usergrant.UserGrantAddedEvent)
+	if !ok {
+		logging.LogWithFields("PROJE-WYOHD", "seq", event.Sequence(), "expectedType", usergrant.UserGrantAddedType).Error("wrong event type")
+		return nil, errors.ThrowInvalidArgument(nil, "PROJE-MQHVB", "reduce.wrong.event.type")
+	}
+	return crdb.NewCreateStatement(
+		e,
+		[]handler.Column{
+			handler.NewCol(UserGrantID, e.Aggregate().ID),
+			handler.NewCol(UserGrantResourceOwner, e.Aggregate().ResourceOwner),
+			handler.NewCol(UserGrantCreationDate, e.CreationDate()),
+			handler.NewCol(UserGrantChangeDate, e.CreationDate()),
+			handler.NewCol(UserGrantSequence, e.Sequence()),
+			handler.NewCol(UserGrantUserID, e.UserID),
+			handler.NewCol(UserGrantProjectID, e.ProjectID),
+			handler.NewCol(UserGrantGrantID, e.ProjectGrantID),
+			handler.NewCol(UserGrantRoles, pq.StringArray(e.RoleKeys)),
+			handler.NewCol(UserGrantState, domain.UserGrantStateActive),
+		},
+	), nil
+}
+
+func (p *UserGrantProjection) reduceChanged(event eventstore.Event) (*handler.Statement, error) {
+	var roles pq.StringArray
+
+	switch e := event.(type) {
+	case *usergrant.UserGrantChangedEvent:
+		roles = e.RoleKeys
+	case *usergrant.UserGrantCascadeChangedEvent:
+		roles = e.RoleKeys
+	default:
+		logging.LogWithFields("PROJE-dIflx", "seq", event.Sequence(), "expectedTypes", []eventstore.EventType{usergrant.UserGrantChangedType, usergrant.UserGrantCascadeChangedType}).Error("wrong event type")
+		return nil, errors.ThrowInvalidArgument(nil, "PROJE-hOr1E", "reduce.wrong.event.type")
+	}
+
+	return crdb.NewUpdateStatement(
+		event,
+		[]handler.Column{
+			handler.NewCol(UserGrantChangeDate, event.CreationDate()),
+			handler.NewCol(UserGrantRoles, roles),
+			handler.NewCol(UserGrantSequence, event.Sequence()),
+		},
+		[]handler.Condition{
+			handler.NewCond(UserGrantID, event.Aggregate().ID),
+		},
+	), nil
+}
+
+func (p *UserGrantProjection) reduceRemoved(event eventstore.Event) (*handler.Statement, error) {
+	switch event.(type) {
+	case *usergrant.UserGrantRemovedEvent, *usergrant.UserGrantCascadeRemovedEvent:
+		// ok
```

Comment:
what does this //ok mean? Event should be handled?

Reply:
events will be handled... it's just the easiest way to catch both event types and remove events don't have any data